server/WFCORE-4291_restore_legacy_not_graceful_startup_mode.adoc

```diff
@@ -0,0 +1,69 @@
+= WFCORE-4291 - Restore legacy (not "graceful") startup mode
+:author:            Jason Lee
+:email:             jasondlee@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+Since Wildfly 11, the server starts up in suspended mode, thus rejecting all incoming requests, until the server has completed booting, at which point the state is set to `RUNNING` and requests are accepted. While this helps ensure that the system is in a stable state before allowing any work to happen, it does have a negative effect on some deployment scenarios reported by Wildfly users. 
+
+One example scenario is this: if Deployment A depends -- at startup -- on external configuration, fetched by an HTTP request to Deployment B, Deployment A will fail to start as the HTTP request for that configuration will be rejected by the starting server.
+
+This RFE proposes a new startup mode, `graceless`, that will allow -- for some subsystems, at least -- requests to be handled much earlier in the application server boot process. In standalone mode, the system administrator can choose to enable this startup mode for that server, and in domain mode, the server group can be configured similarly, applying the configuration choice to all servers in that domain.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFCORE-4291[WFCORE-4291]
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/EAP7-1320[EAP7-1320]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Testing By
+* [X] QE
+
+=== Affected Projects or Components
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+For standalone mode, this option will be exposed as a command line option only, with the addition of a `--graceless` option. Since the value is needed before standalone.xml is parsed, it must come from the command line. Given this lack of start mode persistence, system administrators managing standalone instances will need to update their system scripts (`/etc/init.d`, `sysvinit`, `systemd`, etc.)
```

Comment:
The standalone server has a `start-mode` argument. Its value can be `normal`, `suspend`, `admin-only`.

How the `--graceless` argument will work in combination with the possible values of `start-mode`?

My understanding is:

- start-mode=normal --graceless --> The affected subsystems will process requests when the server is starting.
- start-mode=suspend --graceless --> The affected subsystems will be suspended, I understand requests would be rejected even when the server is starting. 
- start-mode=admin-only --graceless --> When the server starts in the admin-only mode it always rejects requests, so  I understand requests will be rejected even when the server is starting.

If the above statements are the desired behavior when those arguments are used in combination, should not `--graceless` be a `start-mode` value instead of a new argument?

Reply:
Brian and I discussed that at length and felt that adding a new value for start-mode was unwarranted, as it greatly increases the complexity in all of the state transitions for what is a simple corner case. We decided, then, to add a new, separate CLI switch to enable this behavior, which is subject to the start-mode interactions you detailed above.